### ISSUE TITLE
allow comments of some sort in jexl

### ISSUE THREAD

**Issue by gregglind** [mozilla/normandy#861]
Per previous discussions, allowing 'comments' would help a lot.

I evaluated several paths:
1 upgrading `jexl` to support comments of this form '//' and '/* */'
2 pre-stripping WHOLE LINE comments ONLY `// this is a comment` or `# comment`

I think the *dirtiest quickest way* is to do 2 above, maybe using `#` as the character.  

Claims and support: 
- Doing whole line strip is pretty safe.  
- Whole line comments are better for readability and style.  

Rejecting the 'fix jexl' option.  In jexl, the lexer / parser are all hand-written and idiosyncratic.  A fully complete and robust 'inline' comments lexer / parser has a lots of edge cases  "This is my /* string */", and such.  It might all just work though :) 

Example Lexing:
```
> L2.tokenize("a == '/* comment \n blah */' /* a comment (b == 2) /* */")
[ { type: 'identifier', value: 'a', raw: 'a ' },
  { type: 'binaryOp', value: '==', raw: '== ' },
  { type: 'literal',
    value: '/* comment \n blah */',
    raw: '\'/* comment \n blah */\' ' },
  { type: 'openComment', value: '/*', raw: '/* ' },
  { type: 'identifier', value: 'a', raw: 'a ' },
  { type: 'identifier', value: 'comment', raw: 'comment ' },
  { type: 'openParen', value: '(', raw: '(' },
  { type: 'identifier', value: 'b', raw: 'b ' },
  { type: 'binaryOp', value: '==', raw: '== ' },
  { type: 'literal', value: 2, raw: '2' },
  { type: 'closeParen', value: ')', raw: ') ' },
  { type: 'openComment', value: '/*', raw: '/* ' },
  { type: 'closeComment', value: '*/', raw: '*/' } ]
```

(This supports 'example filters', and a bunch of other pieces for recipe creation smoothing)

**Comment by mythmon** [comment]
Thanks for this issue @gregglind. I home we'll be able to add this directly to our fork of jexl, instead of hacking it into normandy. We might still just do the "until the end of line" thing, but it will be in the library.

Also, can you expand out your code example there? I tried to replicate it, but I don't know what `L2` is. Can you include a full script including `require` calls?

**Comment by gregglind** [comment]
So I did a LOT MORE WORK ON THIS.

What does #869 do?
1.  In the `Lexer` catches OCTOTHORPE comments using a REGEX, and DROPS those tokens.  Thus the parser never has to see them, and never has to do anything.
2.  If we can write a regex for C-comments, we could do the same thing.  PS, those comments are much harder to regex, and I gave up with trying to get the right level of slash escaping to work :) , but it's easy to extend.
3.  PARSING the C-Comments is a LOT MORE CONFUSING for me.  I tried writing some grammar above, but I got totally lost in `states.js`.
4.  PS, I don't know how you are handling multi-line.  I wrote a test for that in the jexl, and it fell-apart.
